README.md
What is this?
-------------
A script which produces stylable scrollbars. It also allows styling of the resize handle.
[Check out the demo page](https://desto-git.github.io/scrollbar.js/demo/)



Aren't there a ton of those already?
------------------------------------
Yes, but they all had at least 1 of the following shortcomings:
- Dependent on a library or framework
- Only works vertically
- Native scroll behavior is lost (hello autoscrolling) or messily emulated (e.g. smooth scrolling isn't applied)
- Doesn't dynamically adjust when content changes
- Messes with the element (nesting content into a new div, applying css rules, ...)
- Requires setting height/width on the element to work
- Probably some other things I forgot before writing it down

However, I can't blame them. I couldn't come up with a solution that doesn't have any of these problems either. See the [Limitations](#limitations) and [What have I tried?](#what-have-i-tried) section below for more on that topic.



Browser support?
----------------
I've only tested it in Internet Explorer 11, Firefox 47 and Chrome 51. It probably works on older versions as well.



Goals
-----
I only want to style scrollbars, so...
- It should behave like the native one
- It shouldn't mess with existing elements



Limitations
-----------
- Since there is no way to figure out the distance the native scrollbar would scroll on click, the distance has to be guessed (see `tools/scrollDistance/`)
- The content of the target element will be placed inside another div. That's to retain native scrolling behavior with the mouse and keyboard. If you dynamically add content to it, check if the class `scrollbarjs` is present to add it to its first child instead.
- Doesn't work with textareas (yet?)
- Doesn't work with direction:rtl (yet?)
- `padding-right` doesn't properly work on containers that overflow horizontally. That's not a problem with this script though. This behavior already exists in natively overlflowing containers.
- While modifying the scrollbar of the body does work, I would suggest against doing so on mobile devices. The autohide behavior of the url bar will break. It would hide only after scrolling all the way to the bottom, and then a bit. Same on the way back up, which is really frustrating.



Installation
------------
Just include the script, call `scrollbarjs.init()` and add elements via `scrollbarjs.add( element )`.

```html
<!-- at the end of body -->
<script src="scrollbar.js"></script>
<script>
	scrollbarjs.init();
	var $elems = document.getElementsByClassName('custom-scrollbar');
	for( var i = 0; i < $elems.length; ++i ){
		scrollbarjs.add( $elems[i] );
	}
</script>
</body>
```

You might also want to link `src/scrollbar-basic.css` when starting out. This file contains some basic styling.

```html
<link rel="stylesheet" href="scrollbar-basic.css">
```

You can change some default values by passing an object to `scrollbarjs.init()`.
- `prefix: string = "scrollbarjs"` the prefix to use for generated classes (don't forget to change the .css accordingly)
- `buttonDistance: number = 60` the distance to scroll when clicking on a scrollbar button.
	If this is an integer, it will scroll by that many pixels.
	If this is a floating point number, it will be used as a multiplicator (e.g. 0.90 scrolls by 90%). If you want to scroll by 100%, you have to specifiy something really close to it, like 0.9999. That's because JS can't actually differentiate between ints and floats.
- `trackDistance: number = 0.9` the distance to scroll when clicking on the track (the part between button and thumb).
	Values act the same as above.
- `delay: number = 200` the time in milliseconds to wait until the mousedown event repeats
- `repeat: number = 50` the time in milliseconds to wait between repeats

`scrollbarjs.init()` returns an object containing some information that you might want to make use of:
- `nativeHeight: number` height of the native scrollbar
- `nativeWidth: number` width of the native scrollbar
- `nativeDisplaces: boolean` does the native scrollbar displace the content? (usually desktop browsers do, mobile ones don't)
- `supportsWebkitStyling: boolean` are `::-webkit-scrollbar` pseudo classes supported? [WebKit Blog](https://webkit.org/blog/363/styling-scrollbars/)
- `supportsMsHiding: boolean` is `-ms-overflow-style` supported? [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/-ms-overflow-style)
- `supportsScrollbarWidth: boolean` is `scrollbar-width` supported? [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-width)
- `supportsNativeHiding: boolean` is any of the above supported? They all explicitly check for hiding the native scrollbar.

```javascript
var scrollbarProperties = scrollbarjs.init({
	prefix: 'sb',
	buttonDistance: 0.25
});
```



Note!
-----
- If you have padding on a target element, apply it to `target > .scrollbarjs-viewport` instead.
- On the body element, the body's height will be set to 100vh. Otherwise, it would gradually grow when it's smaller, or it won't scroll if it is bigger.


How does it work?
-----------------
First the script will figure out the native scrollbar width and height. It does so by creating an element with scrollbars and calculating `offsetWidth - clientWidth` and the same for height. This element will be removed afterwards.

`scrollbarjs.init()` will add some required CSS styling to the page, as well as adjust some variables, if a fitting object is passed as a paramater.

When adding an element (`$elem`) via `scrollbarjs.add($elem)`, its content will be copied to a newly created div (`$viewport`). `$elem`'s scrollbars will be hidden via the CSS rule `overflow:hidden;` while `$viewport` will have a negative margin to hide its native scrollbars. This way the mousewheel, arrow keys and other means of scrolling still work like normal. `$viewport` will be placed inside `$elem`, alongside the scrollbars.

Using `requestAnimationFrame()`, the script will check each frame if a scrollbar or the `$viewport` itself needs to be updated. Therefore the size and scroll position of the `$viewport` will be saved to its dataset. If it changes, the scrollbars will be updated.



What have I tried?
------------------
Just putting `overflow:hidden;`:
- Native scroll behavior is lost
- It's still possible to scroll via JavaScript, but that might feel off depending on the default browser scrolling

Put scrollbar on top of the native one:
- Maybe I want a narrow scrollbar? The native one won't be fully covered then
- Maybe I want a transparent scrollbar? The native one would shine through then

Place absolutely positioned div above the target area:
- Well, scrolling works, but selecting text is no longer possible
- Adding mousedown event listener to remove the "overlay" didn't work as hoped (text selection still was dodgy)
- Also having to deal with passing events to the actual element was too fiddly

Put the content into a nested div:
- This is the one I settled with, as it feels the most natural, at the expense of more work to integrate it into ones website
- Messes with the content, so if content is dynamically added, extra care must be taken
- Eventually conflicts with existing CSS rules

Wrap the entire element into a new div:
- a bunch of CSS rules need to be copied (margin, positon, display border, border-radius, outside box-shadow, width, max-width, ...) to the wrapper...
- ...and then be removed from the element to not make things look weird
- watching for CSS rule changes (via adding a class, using developer tools, ...) might be a CPU-intensive task compared to just checking the size and scroll position


CHANGELOG
1.4.0
-----
- throw error when calling init multiple times
- throw error when adding elements before having called init before
- reduce file size even further (7kb -> 6.2kb)
- check for scrollbar-width css support on init

1.3.2
-----
- reduce file size of scrollbar.min.js (8kb -> 7kb)
- forget to bump version in scrollbar.min.js copyright comment, oops!

1.3.1
-----
- make project an npm package
- change license to ISC, which is basically the same as MIT

1.3.0
-----
- scrollbarjs.init() now returns an object with various metadata about the browsers scrolling properties

1.2.0
-----
- using scrollbarjs.add() before scrollbarjs.init() has been called doesn't do anything
- add class "scrollbarjs-active" to the element while a thumb is being dragged
- scrollbarjs.init() can only be called once
- fixed click repeat (holding mouse button) now working

1.1.0
-----
- Consistent line end character across all files (LF)
- add class "scrollbarjs-button-inactive" to buttons when the viewport can't scroll further in that direction
- fixed horizontal thumb being offset by 1px to the top in Chrome in scrollbarjs-basic.css

1.0.0
-----
- Initial release

demo/script.js
var paragraphs = [
	'Magni veniam iste aut nemo. Id atque quia cum totam dolore velit. Rerum dignissimos modi et alias non aliquid dolorem. Est perspiciatis commodi aut accusamus eos maxime itaque quos.<br>Quos delectus ipsam dicta voluptas quo earum aut inventore. Modi sequi quo excepturi recusandae minima aut corrupti.',
	'Et aut reprehenderit blanditiis est eum ratione inventore harum. Eaque nihil quis ut veniam fugiat. Iste inventore voluptatum rerum eius nesciunt saepe.<br>Et et mollitia aliquam quaerat. Minus iusto voluptatem possimus.',
	'Odit quis cupiditate aut. Nisi quam est est officia. Quo dolor laboriosam excepturi iure cum harum eaque et.<br>Dolorem vitae accusamus et enim eaque minima nihil. Saepe minima quia libero beatae labore consequatur dicta omnis. Veritatis ut aut id exercitationem molestiae.',
	'Suscipit soluta quia magni. Eum ipsa architecto aut facilis est rerum dolorem quas. Totam perferendis tenetur vel minima odio autem magni non.<br>Officia in voluptatibus facere tempore voluptatem aut odio nam. Ipsum voluptatibus excepturi sint quis et molestias. Ab consequatur enim eos nostrum ullam.',
	'Rerum sit veniam sapiente. Dolores sequi rem accusantium.<br>Perspiciatis ut ratione modi ut ut et consequatur. Iusto sit cupiditate suscipit temporibus officia. Autem aspernatur odit impedit enim doloremque ea.'
];
var pIndex = 0;

function addP( $target, amount ){
	for( var i = 0; i < amount; ++i ){
		$target.innerHTML += '<p class="lorem">' + paragraphs[ pIndex ] + '</p>';

		++pIndex;
		pIndex %= paragraphs.length;
	}
}

var $content = document.getElementsByClassName('content');
for( var i = 0; i < $content.length; ++i ){
	addP( $content[i], paragraphs.length );
}



// link issues
var $a = document.getElementsByTagName('a');
for( var i = 0; i < $a.length; ++i ){
	if( $a[i].innerHTML.indexOf('#') === 0 )
		$a[i].href = 'https://github.com/desto-git/scrollbar.js/issues/' + $a[i].innerHTML.substr(1);
}



// theme selection
var $theme = document.getElementById('theme');
var $sbjs = document.getElementsByClassName('sbjs');

function selectTheme( themeName ){
	$theme.href = '../src/scrollbar-' + themeName + '.css';
}

$theme.addEventListener('load', function(){
	// when elements are added to scrollbarjs before styling is applied, the calculation of the thumb height cannot work correctly
	// to help this, an update needs to be triggered. this is done by just scrolling by one pixel in this case
	for( var i = 0; i < $sbjs.length; ++i ){
		if( !$sbjs[i].classList.contains('scrollbarjs') ) continue;

		var $target = $sbjs[i].children[0];
		var scrollTop = $target.scrollTop;
		console.log(scrollTop);
		// add some delay, otherwise the change happens too fast for the update to notice
		$target.scrollTop += 1;
		requestAnimationFrame(function(){
			$target.scrollTop -= 1; // in case the container was already scrolled to the bottom
			requestAnimationFrame(function(){
				$target.scrollTop = scrollTop;
			});
		});
	}
});

src/scrollbar.js
var scrollbarjs = (function(){ 'use strict';

/*
	TODO:
	min-width/height on thumbs!!!
	refactor global variables
	textareas
	right to left text
	some way to reliably add padding-right the viewport
	look for a way to get padding-bottom work without the ::after workaround. It just might resolve the padding-right problem as well.
	- `smooth: (boolean) [=true]` should the emulated scrolling be smoothed?
	- `easing: (function) [=easeInOutQuad]` https://gist.github.com/gre/1650294
	probably more

	a more attractive demo site which also explains things in its scrollable boxes
*/

// don't forget to change your css accordingly
var configPrefix = 'scrollbarjs';

// distance to scroll when clicking
// int = px, float = %
// note however that 1.0 === 1, so it will be identified as an int. to say 100%, use something very close to it, e.g. 0.9999
var configButtonDistance = 60; // on a button
var configTrackDistance = 0.9; // somewhere on the track

var configDelay = 200; // wait x milliseconds before repeating the action
var configRepeat = 50; // repeat every x milliseconds



// put some global things into local scope, so uglify can better compress the source
//
var doc = document;
var $head = doc.head;
var $body = doc.body;
var createElement = doc.createElement.bind( doc );

function appendChild( $parent, $child ){ $parent.appendChild( $child ); }
function addEventListener( $target, event, callback ){ $target.addEventListener( event, callback ); }

function addClass( $elem, className ){ $elem.classList.add   ( className ); }
function remClass( $elem, className ){ $elem.classList.remove( className ); }



// getPrefixed( 'viewport', 'fit-parent' ) => "CFG_PREFIX-viewport CFG_PREFIX-fit-parent"
function getPrefixed( /* ... */ ){
	var className = '';

	for( var i = 0; i < arguments.length; ++i )
		className += ' ' + configPrefix + '-' + arguments[i];

	return className.trim(); // remove leading whitespace
}



var INIT = false;
var INIT_DATA = {};
function init( config ){
	if( INIT ){
		console.warn('scrollbarjs cannot be initiated more than once');
		return INIT_DATA;
	}

	if( config !== undefined ){
		if( config.prefix         ) configPrefix          = config.prefix;
		if( config.buttonDistance ) configButtonDistance = config.buttonDistance;
		if( config.trackDistance  ) configTrackDistance  = config.trackDistance;
		if( config.delay          ) configDelay           = config.delay;
		if( config.repeat         ) configRepeat          = config.repeat;
	}

	// inject required CSS into the page
	var $style = createElement('style');
	$style.innerHTML =
		'body.' + configPrefix + '{' +
			'height:100vh;' + // min-height would gradually grow, max-height wouldn't show a scrollbar
		'}' +

		'.' + getPrefixed('drag') + '{' + // don't trigger selection when dragging the thumbs - https://stackoverflow.com/a/4407335
			'-webkit-touch-callout:none;' +
			  '-webkit-user-select:none;' +
			   '-khtml-user-select:none;' +
			     '-moz-user-select:none;' +
			      '-ms-user-select:none;' +
			          'user-select:none;' +
		'}' +

		'.' + configPrefix + '{' +
			'overflow:hidden;' +
			'position:relative;' +
		'}' +

		'.' + getPrefixed('viewport') + '{' +
			'overflow:scroll;' +
			'margin-right:'  + -NATIVE_WIDTH  + 'px;' +
			'margin-bottom:' + -NATIVE_HEIGHT + 'px;' +
		'}' +


		'.' + getPrefixed('resizable') + '{' + // resizing is done in js
			'resize:none!important;' +
		'}';
	appendChild( $head, $style );

	INIT = true;
	INIT_DATA = {
		nativeHeight          : NATIVE_HEIGHT,
		nativeWidth           : NATIVE_WIDTH,
		nativeDisplaces       : ( NATIVE_HEIGHT > 0 || NATIVE_WIDTH > 0 ),
		supportsWebkitStyling : SUPPORTS_WEBKIT_STYLING,
		supportsMsHiding      : SUPPORTS_MS_HIDING,
	};

	return INIT_DATA;
}



// below this point is calculated stuff

// to get the native width and height of a scrollbar, we create an element with overflow:scroll
// it will be removed right afterwards when we have the details we want
var NATIVE_WIDTH  = 0;
var NATIVE_HEIGHT = 0;

// additionally, we check for some ways to hide the scrollbar natively (without killing scrolling like overflow:hidden does)
// in these cases, putting a wrapper around to hide the scrollbars would not be required
var SUPPORTS_WEBKIT_STYLING = false; // do ::-webkit-scrollbar pseudo classes work? note that these don't seem to work on the body in mobile view
var SUPPORTS_MS_HIDING = false; // does -ms-overflow-style work?

(function(){ // scoping to pretent it didn't happen

	// add styles required for this to work
	var $style = createElement('style');
	appendChild( $head, $style );
	$style.innerHTML = '#scrollbarjs-sb-size{' +
		'width:100%;'        + // it needs to be bigger than the scrollbar
		'height:100%;'       +
		'overflow:scroll;'   + // show scrollbars
		'position:absolute;' + // don't break the flow of the document in the process
		'top:100%;'          + // hide
	'}';

	// get dimensions of native scrollbars
	var $size = createElement('div');
	$size.id = 'scrollbarjs-sb-size';
	appendChild( $body, $size );

	NATIVE_WIDTH  = $size.offsetWidth  - $size.clientWidth;
	NATIVE_HEIGHT = $size.offsetHeight - $size.clientHeight;

	// check if native tricks are available to hide scrollbars
	$style.innerHTML +=
		'#scrollbarjs-sb-size::-webkit-scrollbar{display:none}' + // webkit
		'#scrollbarjs-sb-size{-ms-overflow-style:none}'; // ie and edge

	try{
		if( getComputedStyle( $size, '::-webkit-scrollbar' )['display'] === 'none' ){
			SUPPORTS_WEBKIT_STYLING = true;
		}
	}catch(e){
		// Firefox throws when checking an unknown pseudo element (NS_ERROR_NOT_AVAILABLE)
	}

	if( getComputedStyle( $size )['-ms-overflow-style'] === 'none' ){
		SUPPORTS_MS_HIDING = true;
	}

	$style.parentElement.removeChild( $style );
	$size .parentElement.removeChild( $size  );
}());

var TID; // contains the timeout id so we can stop it
var IID; // interval id

var DRAG_MODE; // DRAG_SCROLL or DRAG_RESIZE
var DRAG_AXIS; // scroll vertically or horizontally?
var DRAG_DISTANCE; // distance to scroll in pixels for each pixel the thumb has been moved
var $DRAG_TARGET; // the content being scrolled
var VIEWPORT_SCROLLPOS; // the current x or y position as a double instead of an int, used to circumvent unprecise dragging

function reset(){
	clearInterval( IID );
	clearTimeout(  TID );
	DRAG_MODE = undefined;
	remClass( $body, getPrefixed('drag') );
	if( $DRAG_TARGET !== undefined ){
		remClass( $DRAG_TARGET.parentElement, getPrefixed('active') );
		$DRAG_TARGET = undefined;
	}
}
addEventListener( $body, 'mouseup', reset );
addEventListener( $body, 'mouseenter', function( e ){
	// leaving browser window and going back in, check if left mouse button is "still" pressed
	if( ( e.buttons & 1 ) !== 1 ) reset();
} );

var LAST_X, LAST_Y; // last x y offset
var DRAG_MODE_SCROLL = 0;
var DRAG_MODE_RESIZE = 1;
addEventListener( $body, 'mousemove', function( e ){
	if( DRAG_MODE === DRAG_MODE_SCROLL ){

		if( DRAG_AXIS === DRAG_VER ){
			VIEWPORT_SCROLLPOS += DRAG_DISTANCE * ( e.clientY - LAST_Y );
			scrollViewport( $DRAG_TARGET, null, Math.round( VIEWPORT_SCROLLPOS ), true );

		} else {
			VIEWPORT_SCROLLPOS += DRAG_DISTANCE * ( e.clientX - LAST_X );
			scrollViewport( $DRAG_TARGET, Math.round( VIEWPORT_SCROLLPOS ), null, true );
		}

	} else if( DRAG_MODE === DRAG_MODE_RESIZE ){
		// inconsistent behavior between browsers when resizing centered content
		// chrome's way makes more sense (resize by twice the amount when centered so it keeps following the cursor properly)
		// but firefox' is easier to implement, as I don't need to check for all the ways the content could be centered (margin: 0 auto, flexbox, text-align,..?)
		// and IE11 doesn't understand resize in the first place, so only benefits there!
		var x = ( e.clientX - LAST_X );
		var y = ( e.clientY - LAST_Y );

		if( DRAG_AXIS !== DRAG_HOR  ){ // VER or BOTH
			var h = $DRAG_TARGET.clientHeight;
			$DRAG_TARGET.style.height = h + y + 'px';
			// verify if height is legit, e.g. in case max-height is set
			$DRAG_TARGET.style.height = getComputedStyle( $DRAG_TARGET )['height'];
		}

		if( DRAG_AXIS !== DRAG_VER  ){ // HOR or BOTH
			var w = $DRAG_TARGET.clientWidth;
			$DRAG_TARGET.style.width = w + x + 'px';
			$DRAG_TARGET.style.width = getComputedStyle( $DRAG_TARGET )['width'];
		}
	}

	LAST_X = e.clientX;
	LAST_Y = e.clientY;
} );



// event listeners
var BUTTON_UP    = 0;
var BUTTON_RIGHT = 1;
var BUTTON_DOWN  = 2;
var BUTTON_LEFT  = 3;
function clickButton( event, $viewport, direction ){
	if( event.button !== 0 ) return; // only work on left click

	switch( direction ){
		case BUTTON_UP:    scrollViewport( $viewport, 0, -configButtonDistance    ); break;
		case BUTTON_RIGHT: scrollViewport( $viewport,     configButtonDistance, 0 ); break;
		case BUTTON_DOWN:  scrollViewport( $viewport, 0,  configButtonDistance    ); break;
		case BUTTON_LEFT:  scrollViewport( $viewport,    -configButtonDistance, 0 ); break;
	}
}

var TRACK_UP    = 0;
var TRACK_RIGHT = 1;
var TRACK_DOWN  = 2;
var TRACK_LEFT  = 3;
function clickTrack( event, $viewport, direction ){
	if( event.button !== 0 ) return;

	switch( direction ){
		case TRACK_UP:    scrollViewport( $viewport, 0, -configTrackDistance ); break;
		case TRACK_RIGHT: scrollViewport( $viewport, configTrackDistance, 0  ); break;
		case TRACK_DOWN:  scrollViewport( $viewport, 0, configTrackDistance  ); break;
		case TRACK_LEFT:  scrollViewport( $viewport, -configTrackDistance, 0 ); break;
	}
}

var DRAG_VER = 0;
var DRAG_HOR = 1;
function clickThumb( event, $viewport, direction ){
	if( event.button !== 0 ) return;
	addClass( $body, getPrefixed('drag') );
	addClass( $viewport.parentElement, getPrefixed('active') );

	var $target = event.target.parentElement;
	DRAG_MODE = DRAG_MODE_SCROLL;
	DRAG_AXIS = direction;
	$DRAG_TARGET = $viewport;

	if( direction === DRAG_VER ){
		VIEWPORT_SCROLLPOS = $viewport.scrollTop;
		DRAG_DISTANCE = $viewport.scrollHeight / $target.clientHeight;
	} else {
		VIEWPORT_SCROLLPOS = $viewport.scrollLeft;
		DRAG_DISTANCE = $viewport.scrollWidth / $target.clientWidth;
	}
}

var DRAG_BOTH = 2;
function clickResize( event, $elem, direction ){
	if( event.button !== 0 ) return;
	addClass( $body, getPrefixed('drag') );

	DRAG_MODE = DRAG_MODE_RESIZE;
	DRAG_AXIS = direction;
	$DRAG_TARGET = $elem;
}

// helper functions
var CLASS_SCROLL_NONE = getPrefixed('scroll-none');
var CLASS_SCROLL_VER  = getPrefixed('scroll-ver' );
var CLASS_SCROLL_HOR  = getPrefixed('scroll-hor' );
var CLASS_SCROLL_BOTH = getPrefixed('scroll-both');
function switchScrollClass( $elem, className ){
	if( $elem.classList.contains( className ) )
		return;

	remClass( $elem, CLASS_SCROLL_NONE );
	remClass( $elem, CLASS_SCROLL_VER  );
	remClass( $elem, CLASS_SCROLL_HOR  );
	remClass( $elem, CLASS_SCROLL_BOTH );
	addClass( $elem, className );
}

function scrollViewport( $viewport, x, y, jump, repeat ){
	if( jump === undefined ) jump = false;
	if( repeat === undefined ) repeat = false;

	if( jump ){
		if( x !== null ) $viewport.scrollLeft = x;
		if( y !== null ) $viewport.scrollTop  = y;
	} else {
		if( x % 1 === 0 ) $viewport.scrollLeft += x;
		else              $viewport.scrollLeft += $viewport.clientWidth  * x;
		if( y % 1 === 0 ) $viewport.scrollTop  += y;
		else              $viewport.scrollTop  += $viewport.clientHeight * y;
	}

	if( !repeat && !jump ){ // there's no need to repeat when jumping
		TID = setTimeout( function(){
			IID = setInterval( function(){
				scrollViewport( $viewport, x, y, jump, true );
			}, configRepeat );
		}, configDelay - configRepeat );
	}
}



// contains all the containers to watch over so we don't have to read the entire DOM every time update is called just to find our elements
var $ELEMS = [];

function add( $elem ){
	if( !INIT ) return; // do nothing if scrollbarjs hasn't been initialized yet; that is to prevent weird behavior caused by missing CSS rules

	var $viewport;

	if( $elem.tagName === 'BODY' ){
		$viewport = createElement('div');

	} else if( $elem.tagName === 'TEXTAREA' ){
		// TODO: make textareas work
		$viewport = createElement('div');

	} else {
		$viewport = createElement( $elem.tagName );
	}

	addClass( $elem, configPrefix );

	$viewport.className = getPrefixed('viewport');
	$viewport.dataset.scrollbarjs = ' '; // a space so I don't have to do additional checks in the update function
	$viewport.innerHTML = $elem.innerHTML;

	$elem.innerHTML = '';
	appendChild( $elem, $viewport );

	// create scrollbars; naming similar to https://webkit.org/blog/363/styling-scrollbars/
	// vertical
	var $ver = createElement('aside');
	$ver.className = getPrefixed('scrollbar', 'ver');

	var $up = createElement('div');
	$up.className = getPrefixed('button', 'up');
	addEventListener( $up, 'mousedown', function( e ){ clickButton( e, $viewport, BUTTON_UP ) } );

	var $trackVer = createElement('div');
	$trackVer.className = getPrefixed('track', 'track-ver');

	var $trackUp = createElement('div');
	$trackUp.className = getPrefixed('track-piece', 'track-up');
	addEventListener( $trackUp, 'mousedown', function( e ){ clickTrack( e, $viewport, TRACK_UP ) } );

	var $thumbVer = createElement('div');
	$thumbVer.className = getPrefixed('thumb', 'thumb-ver');
	addEventListener( $thumbVer, 'mousedown', function( e ){ clickThumb( e, $viewport, DRAG_VER ) } );

	var $trackDown = createElement('div');
	$trackDown.className = getPrefixed('track-piece', 'track-down');
	addEventListener( $trackDown, 'mousedown', function( e ){ clickTrack( e, $viewport, TRACK_DOWN ) } );

	appendChild( $trackVer, $trackUp   );
	appendChild( $trackVer, $thumbVer  );
	appendChild( $trackVer, $trackDown );

	var $down = createElement('div');
	$down.className = getPrefixed('button', 'down');
	addEventListener( $down, 'mousedown', function( e ){ clickButton( e, $viewport, BUTTON_DOWN ) } );

	appendChild( $ver, $up       );
	appendChild( $ver, $trackVer );
	appendChild( $ver, $down     );

	// horizontal
	var $hor = createElement('aside');
	$hor.className = getPrefixed('scrollbar', 'hor');

	var $left = createElement('div');
	$left.className = getPrefixed('button', 'left');
	addEventListener( $left, 'mousedown', function( e ){ clickButton( e, $viewport, BUTTON_LEFT ) } );

	var $trackHor = createElement('div');
	$trackHor.className = getPrefixed('track', 'track-hor');

	var $trackLeft = createElement('div');
	$trackLeft.className = getPrefixed('track-piece', 'track-left');
	addEventListener( $trackLeft, 'mousedown', function( e ){ clickTrack( e, $viewport, TRACK_LEFT ) } );

	var $thumbHor = createElement('div');
	$thumbHor.className = getPrefixed('thumb', 'thumb-hor');
	addEventListener( $thumbHor, 'mousedown', function( e ){ clickThumb( e, $viewport, DRAG_HOR ) } );

	var $trackRight = createElement('div');
	$trackRight.className = getPrefixed('track-piece', 'track-right');
	addEventListener( $trackRight, 'mousedown', function( e ){ clickTrack( e, $viewport, TRACK_RIGHT ) } );

	appendChild( $trackHor, $trackLeft  );
	appendChild( $trackHor, $thumbHor   );
	appendChild( $trackHor, $trackRight );

	var $right = createElement('div');
	$right.className = getPrefixed('button', 'right');
	addEventListener( $right, 'mousedown', function( e ){ clickButton( e, $viewport, BUTTON_RIGHT ) } );

	appendChild( $hor, $left     );
	appendChild( $hor, $trackHor );
	appendChild( $hor, $right    );

	appendChild( $elem, $ver );
	appendChild( $elem, $hor );

	// corner
	var $corner = createElement('aside');
	$corner.className = getPrefixed('corner');

	var resize = getComputedStyle( $elem )['resize'];
	if( resize === 'both' || resize === 'horizontal' || resize === 'vertical' ){
		addClass( $elem, getPrefixed('resizable') );

		var direction;
		var directionClass;
		switch( resize ){
			case 'vertical':   direction = DRAG_VER;  directionClass = 'ver';  break;
			case 'horizontal': direction = DRAG_HOR;  directionClass = 'hor';  break;
			case 'both':       direction = DRAG_BOTH; directionClass = 'both'; break;
		}

		addClass( $corner, getPrefixed('resize-' + directionClass) );
		addEventListener( $corner, 'mousedown', function( e ){ clickResize( e, $elem, direction ) } );
	}

	appendChild( $elem, $corner );



	$ELEMS.push( $elem );
}



function updateViewport( $elem, $viewport ){
	// height:100% doesn't work on elements whose parents don't have an explicit value set
	// this is a hacky way to achieve this anyways
	// Firefox won't remember the scroll position after a reload, but I think that's bearable ;)
	if( $elem.clientHeight < $viewport.scrollHeight ){ // check if the element can grow

		var y = $viewport.scrollTop; // keep scrolling distance
		$viewport.style.height = ''; // try to be as big as the content

		if( $elem.clientHeight < $viewport.scrollHeight ) // if that doesn't work, shrink to fit
			$viewport.style.height = $elem.clientHeight + NATIVE_HEIGHT + 1 + 'px'; // +1 so the element will expand once it can, trying this procedure again
		else // if it did work, remove that spare 1px line at the bottom; this will prevent flickering when resizing
			$viewport.style.height = $elem.clientHeight + NATIVE_HEIGHT + 'px';

		$viewport.scrollTop = y;

	} else { // match the element
		$viewport.style.height = $elem.clientHeight + NATIVE_HEIGHT + 'px';
	}
}

function updateScrollbars( $elem, $viewport ){
	// in % from 0 to 1, how much of the viewport is visible
	var thumbHeight = $viewport.clientHeight / $viewport.scrollHeight;
	var thumbWidth  = $viewport.clientWidth  / $viewport.scrollWidth;

	// which scrollbars are active
	     if( thumbHeight === 1 && thumbWidth === 1 ) switchScrollClass( $elem, CLASS_SCROLL_NONE );
	else if( thumbHeight !== 1 && thumbWidth === 1 ) switchScrollClass( $elem, CLASS_SCROLL_VER  );
	else if( thumbHeight === 1 && thumbWidth !== 1 ) switchScrollClass( $elem, CLASS_SCROLL_HOR  );
	else if( thumbHeight !== 1 && thumbWidth !== 1 ) switchScrollClass( $elem, CLASS_SCROLL_BOTH );

	// apply extra classes to indicate if a button will do something
	var $buttonUp    = $elem.children[1].children[0];
	var $buttonDown  = $elem.children[1].children[2];
	var $buttonLeft  = $elem.children[2].children[0];
	var $buttonRight = $elem.children[2].children[2];

	if( $viewport.scrollTop  === 0 ) addClass( $buttonUp,   getPrefixed('button-inactive') );
	else                             remClass( $buttonUp,   getPrefixed('button-inactive') );
	if( $viewport.scrollLeft === 0 ) addClass( $buttonLeft, getPrefixed('button-inactive') );
	else                             remClass( $buttonLeft, getPrefixed('button-inactive') );

	if( ( $viewport.scrollTop  + $viewport.clientHeight ) === $viewport.scrollHeight ) addClass( $buttonDown,  getPrefixed('button-inactive') );
	else                                                                               remClass( $buttonDown,  getPrefixed('button-inactive') );
	if( ( $viewport.scrollLeft + $viewport.clientWidth  ) === $viewport.scrollWidth  ) addClass( $buttonRight, getPrefixed('button-inactive') );
	else                                                                               remClass( $buttonRight, getPrefixed('button-inactive') );

	// calculate thumb size
	var $trackVer  = $elem.children[1].children[1];
	var $trackUp   = $trackVer.children[0];
	var $thumbVer  = $trackVer.children[1];
	var $trackDown = $trackVer.children[2];

	var $trackHor   = $elem.children[2].children[1];
	var $trackLeft  = $trackHor.children[0];
	var $thumbHor   = $trackHor.children[1];
	var $trackRight = $trackHor.children[2];

	var trackHeight = $trackVer.clientHeight;
	var trackWidth  = $trackHor.clientWidth;

	// height/width of track-up/track-left
	var trackUpHeight  = Math.floor( trackHeight * ( $viewport.scrollTop  / $viewport.scrollHeight ) );
	var trackLeftWidth = Math.floor( trackWidth  * ( $viewport.scrollLeft / $viewport.scrollWidth ) );
	$trackUp.style.height  = trackUpHeight  + 'px';
	$trackLeft.style.width = trackLeftWidth + 'px';

	// height/width of the thumb
	var thumbHeight = Math.ceil( trackHeight * thumbHeight );
	var thumbWidth  = Math.ceil( trackWidth  * thumbWidth  );
	$thumbVer.style.height = thumbHeight + 'px';
	$thumbHor.style.width  = thumbWidth  + 'px';

	// height/width of track-down/track-right
	var trackDownHeight = trackHeight - trackUpHeight  - thumbHeight;
	var trackRightWidth = trackWidth  - trackLeftWidth - thumbWidth;
	$trackDown.style.height = trackDownHeight + 'px';
	$trackRight.style.width = trackRightWidth + 'px';
}

function update( $elem ){
	var $viewport = $elem.children[0];

	var viewport  = $elem.clientHeight     + ',' + $elem.clientWidth     + '-' +
	                $viewport.scrollHeight + ',' + $viewport.scrollWidth;
	var scrollbar = $viewport.scrollTop    + ',' + $viewport.scrollLeft;

	var dataOld = $viewport.dataset.scrollbarjs.split(' ');
	if( dataOld[0] !== viewport ){ // element or content size changed
		updateViewport( $elem, $viewport );
		updateScrollbars( $elem, $viewport );

	} else if( dataOld[1] !== scrollbar ) { // only scrollbar position changed
		updateScrollbars( $elem, $viewport );

	} else { // nothing changed, so nothing to do
	 return;
	}

	$viewport.dataset.scrollbarjs = viewport + ' ' + scrollbar;
}

// keep it updated
requestAnimationFrame( function watch(){
	$ELEMS.forEach( update );
	requestAnimationFrame( watch );
} );



return {
	add: add,
	init: init
}

}());
